ring/ring.go

```diff
@@ -760,6 +771,24 @@ func (r *Ring) shuffleShard(identifier string, size int, lookbackPeriod time.Dur
 		return r
 	}
 
+	// If requested shard size covers entire ring, and we don't need to do any filtering of read-only instances,
+	// we can return entire ring directly.
+	if r.readOnlyInstancesUpdated && (size <= 0 || len(r.ringDesc.Ingesters) <= size) {
+		if lookbackPeriod > 0 && (r.readOnlyInstances == 0 || r.oldestReadOnlyUpdatedTimestamp >= lookbackUntil) {
+			return r
+		}
+
+		// for lookbackPeriod == 0, we can only return full ring if there are no read-only instances (no filtering needs to be done),
+		// and all zones have number of instances <= numInstancesPerZone. If the second condition isn't true,
+		// zones are unbalanced and we return full ring, then after adding more instances to the ring, subsequent shuffleShard with lookback
+		// may not return all instances.
+	}
+
+	if size <= 0 {
+		// Use all available instances in each zone.
+		size = math.MaxInt
+	}
+
```

Comment:
I wonder if there's another optimization available:

In the case of `size <= 0` and `lookbackPeriod = 0`, I think we want to return the full ring minus the read-only instances. Rather than falling through to the token looping logic (which will end up looping through ALL the tokens in the ring, since we're looking for `math.MaxInt` total instances), could we instead just loop over the instances themselves, remove the read-only ones, and return the rest?

Reply:
You're right. This is also something that @pracucci suggested. I plan to implement this idea before moving the PR out of draft.

Reply:
I've implemented this idea, and rerun benchmarks. It shows significant improvement for shardSize = 0, as expected.

Reply:
There's another optimization available -- if number of instances in the zone is higher than computed `numInstancesPerZone`, we can also just filter out read-only instances (if needed) and keep all remaining instances from the zone.

Reply:
> There's another optimization available -- if number of instances in the zone is higher than computed `numInstancesPerZone`, we can also just filter out read-only instances (if needed) and keep all remaining instances from the zone.
> 
> I will not implement it in this PR (it's already big enough), but we could give it a try in subsequent PR. However, with caching, benefit will likely be tiny.

I've tried this, and benchmark shows nice improvement. I will send the code change in the next PR.